### PR TITLE
New tests for AeroSurfaces, Environment, Environment Analysis and Function

### PR DESCRIPTION
**Pull request by Gui-FernandesBR**
## Pull request type

- [x] Other (please describe): Tests

## Pull request checklist
- Code base additions (for bug fixes / features):
  - [x] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [ ] All tests (`pytest --runslow`) have passed locally **apperantly not for Environment with --runslow option**

## What is the new behavior?
I'm bringing some new simple tests to the tests folder. 

### New test that were added:
- **test_environment.py**:
    - Now it also tests the methods `info()`, `printEarthDetails()` , `allPlotInfoReturned()`, `allInfoReturned()`, `exportEnvironment()` and `utmToGeodesic` from the Environment class 
- **test_environment_anaysis.py**:
    - This is new! EnvAnalysis was lacking a set of unit tests. The descriptions of what is being tested are very short and simple, but I can describe on next comments if requested.
- **test_function.py**:
    -  Also a new file! I tried to add simple tests to cover only what was not yet being called by other classes. Baby steps for now.
 
All the other changes might desapper when #305 get merged.

### Bugs or other fixes:
- **AeroSurfaces.py**: 
    - There was a wrong `{self.cp:.3f}` format string, which was trying to format a tuple as a float
- **Environment.py** : 
    - Adding `return None` to some methods, which are not only a good practice but also very useful to unit tests
    - Fixing bug when the there's no `atmosphericModelFile` attribute and you try to run the `exportEnvironment()` method
    - Specify the type of input parameters of `utmToGeodesic()`
    - Typo fixes
- **EnvironmentAnalysis.py**:
    - Specify some type of return
    - Returning None on most of the methods

By using the following command line code I was able to track the total number of lines covered by our unit tests.

    coverage run --module pytest --runslow
    ...
    coverage html

First, I ran that before starting to work, then after, and at the end I consolidated the information in the following table. Overall the code coverage has increased 20%

|                   Module                  |   BEFORE   |         |          |    AFTER   |         |          | Improvement |
|:-----------------------------------------:|:----------:|:-------:|:--------:|:----------:|:-------:|:--------:|:-----------:|
|                                           | statements | missing | coverage | statements | missing | coverage |             |
| rocketpy\__init__.py                      |     17     |    0    |   100%   |     17     |    0    |   100%   |      0%     |
| rocketpy\AeroSurfaces.py                  |     337    |   138   |    59%   |     337    |    6    |    98%   |     40%     |
| rocketpy\Environment.py                   |    1182    |   138   |    88%   |    1189    |   367   |    69%   |     -28%    |
| rocketpy\EnvironmentAnalysis.py           |    1174    |   1067  |    9%    |    1193    |    66   |    94%   |     90%     |
| rocketpy\Flight.py                        |    1887    |   251   |    87%   |    1887    |   251   |    87%   |      0%     |
| rocketpy\Function.py                      |     965    |   583   |    40%   |     965    |   416   |    57%   |     30%     |
| rocketpy\Motor.py                         |     371    |   131   |    65%   |     371    |   131   |    65%   |      0%     |
| rocketpy\Parachute.py                     |     20     |    0    |   100%   |     20     |    0    |   100%   |      0%     |
| rocketpy\Rocket.py                        |     228    |    43   |    81%   |     228    |    43   |    81%   |      0%     |
| rocketpy\tools.py                         |     19     |    1    |    95%   |     19     |    1    |    95%   |      0%     |
| rocketpy\units.py                         |     50     |    41   |    18%   |     50     |    18   |    64%   |     72%     |
| rocketpy\utilities.py                     |     54     |    30   |    44%   |     54     |    30   |    44%   |      0%     |
| tests\acceptance\test_bella_lui_rocket.py |     56     |    0    |   100%   |     56     |    0    |   100%   |      0%     |
| tests\acceptance\test_ndrt_2020_rocket.py |     45     |    1    |    98%   |     45     |    1    |    98%   |      0%     |
| tests\conftest.py                         |     33     |    4    |    88%   |     33     |    4    |    88%   |      0%     |
| tests\test_environment_anaysis.py         |      0     |    0    |          |     50     |    0    |   100%   |             |
| tests\test_environment.py                 |     88     |    5    |    94%   |     101    |    15   |    85%   |     -11%    |
| tests\test_flight.py                      |     195    |    0    |   100%   |     195    |    0    |   100%   |      0%     |
| tests\test_function.py                    |      0     |    0    |          |     57     |    0    |   100%   |             |
| tests\test_rocket.py                      |     142    |    6    |    96%   |     149    |    6    |    96%   |      0%     |
| tests\test_solidmotor.py                  |     80     |    0    |   100%   |     80     |    0    |   100%   |      0%     |
| tests\test_utilities.py                   |     20     |    2    |    90%   |     20     |    2    |    90%   |      0%     |
| Total                                     |    6963    |   2441  |    65%   |    7116    |   1357  |    81%   |     20%     |

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by Gui-FernandesBR** [comment]
I also plotted a code coverage x code lines comparison. 
Currently, to achieve 90% cv we need 6404 statements being called by tests. 
A code coverage of 95% would require 6760 statements.



image

File: rocketpy/Environment.py [hunk 1]
```diff
@@ -3329,7 +3330,7 @@ def allInfoReturned(self):
 
     def exportEnvironment(self, filename="environment"):
         """Export important attributes of Environment class so it can be used
-        again in further siulations by using the customAtmosphere atmospheric
+        again in further simulations by using the customAtmosphere atmospheric
         model.
```
Comment (MateusStano) [inline rocketpy/Environment.py:3334]: """
Export important attributes of Environment class to a .json file, saving all the information needed to recreate the same environment using customAtmosphere.
"""

Would be a better description here, might as well change it now.

File: rocketpy/Environment.py [hunk 2]
```diff
@@ -3518,7 +3523,7 @@ def geodesicToUtm(self, lat, lon, datum):
 
         return x, y, utmZone, utmLetter, hemis, EW
 
-    def utmToGeodesic(self, x, y, utmZone, hemis, datum):
+    def utmToGeodesic(self, x: float, y: float, utmZone: int, hemis: str, datum: str):
```
Comment (MateusStano) [inline rocketpy/Environment.py:3526]:
```suggestion
    def utmToGeodesic(self, x, y, utmZone, hemis, datum):
```

Are we gonna start using function annotations now? I thought we had not decided on this yet

File: tests/test_environment.py [hunk 3]
```diff
@@ -160,6 +165,8 @@ def test_era5_atmosphere(mock_show):
 def test_gefs_atmosphere(mock_show, example_env_robust):
     example_env_robust.setAtmosphericModel(type="Ensemble", file="GEFS")
     assert example_env_robust.allInfo() == None
+    assert isinstance(example_env.allPlotInfoReturned(), dict)
+    assert isinstance(example_env.allInfoReturned(), dict)
```
Comment (MateusStano) [inline tests/test_environment.py:169]: The tests just checks if these methods return a dict? Wouldn't it be better to check if the dict has the keys and values that we expect from it?

File: tests/test_environment_anaysis.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        surfaceDataFile="./data/weather/EuroC_single_level_reanalysis_2002_2021.nc",
+        pressureLevelDataFile="./data/weather/EuroC_pressure_levels_reanalysis_2001-2021.nc",
+        timezone=None,
+        unit_system="metric",
+        forecast_date=None,
+        forecast_args=None,
+        maxExpectedAltitude=None,
+    )
+
+    return env
```
Comment (MateusStano) [inline tests/test_environment_anaysis.py:42]: Shouldn't this be in `conftest.py`?

File: tests/test_environment_anaysis.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    assert env.plot_average_surface10m_wind_speed_along_day(False) == None
+    assert env.plot_average_surface10m_wind_speed_along_day(True) == None
+    assert env.plot_average_sustained_surface100m_wind_speed_along_day() == None
+    assert env.plot_average_day_wind_rose_all_hours() == None
+    assert env.plot_average_day_wind_rose_specific_hour(12) == None
+    assert env.plot_average_day_wind_rose_specific_hour(12) == None
+
+
+@patch("matplotlib.pyplot.show")
+def test_profile_plots(mock_show, env):
```
Comment (MateusStano) [inline tests/test_environment_anaysis.py:113]: Missing docstring here

File: tests/test_environment_anaysis.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    assert env.plot_average_wind_heading_profile(clear_range_limits=True) == None
+    assert env.plot_average_wind_speed_profile(clear_range_limits=False) == None
+    assert env.plot_average_wind_speed_profile(clear_range_limits=True) == None
+    assert env.plot_average_pressure_profile(clear_range_limits=False) == None
+    assert env.plot_average_pressure_profile(clear_range_limits=True) == None
+    assert env.plot_wind_profile_over_average_day() == None
+
+
+@patch("matplotlib.pyplot.show")
+def test_animation_plots(mock_show, env):
```
Comment (MateusStano) [inline tests/test_environment_anaysis.py:126]: Missing docstring here

File: tests/test_environment_anaysis.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    assert isinstance(env.animate_wind_gust_distribution_over_average_day(), HTML)
+    assert isinstance(env.animate_wind_heading_profile_over_average_day(), HTML)
+    assert isinstance(env.animate_wind_profile_over_average_day(), HTML)
+    assert isinstance(
+        env.animate_sustained_surface_wind_speed_distribution_over_average_day(),
+        HTML,
+    )
+
+
+def test_exports(env):
```
Comment (MateusStano) [inline tests/test_environment_anaysis.py:139]: Missing docstring here

File: tests/test_function.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+@pytest.fixture
+def func_from_csv():
+    func = Function(
+        source="tests/fixtures/airfoils/e473-10e6-degrees.csv",
+        inputs=["Scalar"],
+        outputs=["Scalar"],
+        interpolation="linear",
+        extrapolation="linear",
+    )
+    return func
```
Comment (MateusStano) [inline tests/test_function.py:36]: I think these should be in `conftest.py` too

**Review by MateusStano — CHANGES_REQUESTED** [review]

**Comment by MateusStano** [comment]
Hey, really good PR!

I think we are still missing docstrings in `test_function.py` too

File: rocketpy/Environment.py (hunk 1 above)
Comment (Gui-FernandesBR) [reply rocketpy/Environment.py]: Fixed

File: rocketpy/Environment.py (hunk 2 above)
Comment (Gui-FernandesBR) [reply rocketpy/Environment.py]: I've inserted this for helping me and forgot to erase, my bad.
solved!

File: tests/test_environment.py (hunk 3 above)
Comment (Gui-FernandesBR) [reply tests/test_environment.py]: Good point, solved.

File: tests/test_environment_anaysis.py (hunk 4 above)
Comment (Gui-FernandesBR) [reply tests/test_environment_anaysis.py]: Yeah! I've just moved everything.

Btw the conftest.py may be starting to become confusing or difficult to read...

**Comment by Gui-FernandesBR** [comment]
@MateusStano thks for the solid comments, now it's back to you for re-review.

Btw you can change the base branch to any other if you think is best

**Comment by MateusStano** [comment]
Great work! I changed some very minor things. The biggest thing was a small bug inside the Function class that returned `nan` when receiving `np.inf` in the `__call__` method in `int` or `float` defined Functions.

I think this ready to merge now.

**Comment by Gui-FernandesBR** [comment]
> Great work! I changed some very minor things. The biggest thing was a small bug inside the Function class that returned `nan` when receiving `np.inf` in the `__call__` method in `int` or `float` defined Functions.
> 
> Another thing, the `EnvironmentAnalysis` tests take veeery long to run. I think we should make them part of the `--run-slow` routine in pytest just like we do with some `Environment` tests. Think this is a good idea?
> 
> If so, I think after doing that we can merge this one.

* Very nice catch! Since the modification at Function.py seems to be a bug, should we commit this directly to beta/v0.1.0?  
Or do you think it is a really minor bug?
* I agree that EnvAnalysis tests might be at the --run-slow option, I'll modify it.
* Could you confirm that tests are passing locally?
* Also, am I pulling it into the correct branch? Just would like to have this double check with you, I'm not sure.

**Comment by MateusStano** [comment]
> * Very nice catch! Since the modification at Function.py seems to be a bug, should we commit this directly to beta/v0.1.0?
>   Or do you think it is a really minor bug?

We could commit this bug fix directly to develop I think. It is a minor bug so I think it is fine the way it is. The reason it started being a problem just now is because of the way `SolidMotor.centerOfMass` is defined, which changed with distances-vs-positions, so it made some static margin test asserts break.

> * Could you confirm that tests are passing locally?

Yeah, everything is running fine

> * Also, am I pulling it into the correct branch? Just would like to have this double check with you, I'm not sure.

You are, this PR has changes that only make sense given the changes made in distances-vs-positions. We would have to change a lot to pull to beta/v1.0

**Comment by Gui-FernandesBR** [comment]
> > * Very nice catch! Since the modification at Function.py seems to be a bug, should we commit this directly to beta/v0.1.0?
> >   Or do you think it is a really minor bug?
> 
> We could commit this bug fix directly to develop I think. It is a minor bug so I think it is fine the way it is. The reason it started being a problem just now is because of the way `SolidMotor.centerOfMass` is defined, which changed with distances-vs-positions, so it made some static margin test asserts break.
> 
> > * Could you confirm that tests are passing locally?
> 
> Yeah, everything is running fine
> 
> > * Also, am I pulling it into the correct branch? Just would like to have this double check with you, I'm not sure.
> 
> You are, this PR has changes that only make sense given the changes made in distances-vs-positions. We would have to change a lot to pull to beta/v1.0

Ok, tks very much. I've set the env analysis tests to slow option. 
Now I'm waiting for your review/approve, please

**Review by MateusStano — APPROVED** [review]

**Comment by MateusStano** [comment]
Good to merge

**Comment by MateusStano** [comment]
Seems #282 is ready to receive this one. Merging now

**Pull request merged**